metricbeat/module/system/cpu/cpu.go

```diff
@@ -55,18 +58,60 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 	return &MetricSet{
 		BaseMetricSet: base,
 		config:        config,
-		cpu:           new(cpu.Monitor),
+		cpu:           metrics.New(""),
 	}, nil
 }
 
 // Fetch fetches CPU metrics from the OS.
 func (m *MetricSet) Fetch(r mb.ReporterV2) error {
-	sample, err := m.cpu.Sample()
+	sample, err := m.cpu.Fetch()
 	if err != nil {
 		return errors.Wrap(err, "failed to fetch CPU times")
 	}
 
-	r.Event(collectCPUMetrics(m.config.Metrics, sample))
+	event := common.MapStr{"cores": sample.CPUCount()}
+
+	for _, selector := range m.config.Metrics {
+		switch strings.ToLower(selector) {
```

Comment:
Maybe this should be an enum?

As the value is parsed from the configuration file, why not do the transformation at init time? Do we have a check that users did configure the setting correctly?

Reply:
So, they're defined as constants, and there's a `validate` method that got dropped somewhere, I'll add it back in.

Reply:
I rather mean something like that:

```
type metricType uint

const (
  metricTypePercentage metricType = iota
  ...
)

func (t metricType) String() string {
   ...
}

func parseMetricType(s string) (metricType, error) {
  ...
}

// parses setting into enum when the module/metricset extracts its config via `cfg.Unpack(&to)`...
func (t *metricType) Unpack(in string) error {
  parsed, err := parseMetricType(in)
  if err == nil {
    *t = parsed
  }
  return err
}
```

Even if the fields `percentages` ... are consts, we should not be required to call `strings.ToLower(...)` or do string comparisona over and over again.

Reply:
So just do a string match and then return the `metricType` iota? That could work, and I agree the `ToLower` is a bit unneeded.

Reply:
yep.